deps: add yarnhook

frosso opened the pull request:
#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

It can happen that we add/remove dependencies between merges.
But it can also happen that not everyone remembers to run `npm install` after new updates.

`yarnhook` detects changes on `package-lock.json` to determine whether or not we need to run `npm install` after checking out a new branch or updating our local branch.

It doesn't mean that it can fail - in my experience it might not run when rebasing or resolving merge conflicts. But it's a helpful tool, I think.

https://github.com/frontsideair/yarnhook

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Check out this branch
* Run `npm install` to get `yarnhook`
* Run `git checkout -b test-deps && npm install --save-dev addition && git add . && git commit -am "test add"` to add a random dependency to a new temp branch
* Run `git checkout frosso/yarnhook`
* You should see the dependencies update running (because of the change of branch)
* Run `git branch -D test-deps` to delete the temp branch

-------------------

- [ ] Added changelog entry (or does not apply)
Don't think it's necessary/worth mentioning publicly, just a dev tool.
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

ismaeldcom reviewed (commented):
Thanks for  proposing this @frosso. I tested it and works as expected, but there is a minor issue. Do you have `yarnhook` installed globally? Because I receive the following error on branch checkout :
```
.husky/post-checkout: line 4: yarnhook: command not found
```

I needed to replace `yarnhook` with `npx --no-install yarnhook` in every husky hook to use the local package.

frosso commented:
@ismaeldcom thanks for checking this out!
You're right, I also had it installed globally - I removed it, checked out this branch and then I was getting the same error as you!
Thanks for pointing it out, I now changed the scripts to `npx` - that way it's not going to error out the first time the changes are checked out 👍

ismaeldcom commented on package.json:
```diff
@@ -56,9 +56,9 @@
-      "post-checkout": "yarnhook",
-      "post-merge": "yarnhook",
-      "post-rewrite": "yarnhook",
+      "post-checkout": "npx --no-install yarnhook",
+      "post-merge": "npx --no-install yarnhook",
+      "post-rewrite": "npx --no-install yarnhook",
```
I think there is no need to use `npx` here, only in the `.husky/*` files. This might be related to a mix of settings for husky v4 and v6. In the v4 you can use local packages directly, but not in the v6.

It is out of the scope of this issue, but we have settings for both versions but miss the `"prepare": "husky install"` part in the scripts section of package to avoid problems with newer versions on npm. Maybe is a good option to create a new issue to make a full migration to v6.

frosso replied:
Thanks for pointing this out! I read about the v4 and v6 migration a bit more because at first I was also confused on the definitions in `package.json`.
I not ran this thing: https://github.com/typicode/husky-4-to-6 (I'm guessing it didn't run in the previous migration).
I don't think it's unreasonable to also migrate in this PR, since you pointed it out and since it's straightforward :) less PRs to review :)

ismaeldcom reviewed (approved):
I just let a little nitpick comment about the npx change 🙂 But other than that, let's 🚢  it!

frosso commented:
@ismaeldcom thanks again for the suggestions and the quick feedback loop! I made some additional changes to fully migrate to husky@v6, cleaning up the config and ensuring that the hooks are not run on GH Actions.

Before my changes the hooks were installed: https://github.com/Automattic/woocommerce-payments/runs/2405382505?check_suite_focus=true#step:5:185
Now they aren't anymore: https://github.com/Automattic/woocommerce-payments/runs/2510403713?check_suite_focus=true#step:5:193

I tried checking out the branch and running `npm ci` to clean up `node_modules` and install the packages.
I am still seeing an error in the console:
> not found: yarnhook

I'm guessing this error will appear to everyone running `git pull` after this branch is merged - but that should be resolved after running `npm install` one last time. I can alert people on Slack when this gets merged 👍

ismaeldcom reviewed (approved):
Thanks @frosso for making the full migration to husky@v6. Better to do it here and not create another PR 🙂 

Good catch with the CI part!

And yes, I think it is normal to receive the `not found` error the first time because we already have husky but not yarnhook. But it's a one time thing.

LGTM, let's 🚢 !
